### DISCUSSION TITLE
Semantic Search API Implementation

### DISCUSSION THREAD

**Discussion by brrksl** [directus/directus#22213] [Feature Requests]
### Summary

Allow for more powerful search operations and behaviors aside from the currently allowed "string and text type fields within a collection." 

### Basic Example

Currently, when using the search query parameter ([https://docs.directus.io/reference/query.html#search](https://docs.directus.io/reference/query.html#search)) you are only allowed to do rudimentary search string/text fields in the root of the collection (so no related item fields). To allow for more flexibility, it would be beneficial to have a more "semantic search" approach. 

Suppose there is are related collections that have this kind of "map":
- Streams (Main collection)
-- Channels (M2M w/ Streams)
-- Shows (M2M w/ Streams)
-- Sports (M2M w/ Streams)
-- Teams (M2M w/ Streams)

So i.e. each stream can have multiple "shows" or "channels" attached to it, so if one is simultaneously casting to multiple channels or if one stream has multiple "segments". 

Assuming "Zone" is referring to a channel name
**https://directus.install/items/Streams?search=Zone**
only searching text fields and returning items in the Streams collection with "Zone" in the title/description, it would be "smarter" and perhaps _behave_ more like this:
- Check title/description for Zone
- Check JSON fields like tags (like using filter operator _in with the returned array from a tags field in the API), 
- Explore relational datasets for term and potential relationships. (i.e. Shows is M2M with channels and "The Zone Show" is on the "Zone" channel, it would identify and take this into consideration).

Then it would return in a form similar to a search engine - by relevance/context (maybe with some rules/boundaries that can be implemented around result sets). 

### Motivation

In my use case, we have user-searchable stream items that have varying text field values. For instance, we have a show called "The Saturday Show" that is on the channel "Zone", but if you go to shows and search for "Zone" nothing will pop up - no relational search. 

Similarly, if you go to one of our streams and search "5a basketball" you get 0 results, however searching "5a boys basketball" returns many results (despite the first making sense to have results for). There is no intelligence toward the text and return values, so user expectations aren't met.

### Detailed Design

I imagine this being a semantic search/vector search implementation. I recently stumbled across ParadeDB ([https://github.com/paradedb/paradedb](https://github.com/paradedb/paradedb)) which is like "open source Elasticsearch" on top of Postgres, interesting project and idea.

By semantic I mean this sort of definition: "Semantic search is a data searching technique which uses the intent and contextual meaning behind a search query to deliver more relevant results." whereas vector search is a sort of implementation methodology for semantic search. More info: [https://www.timescale.com/learn/vector-search-vs-semantic-search](https://www.timescale.com/learn/vector-search-vs-semantic-search).

In Directus, this may look like an "on item.create/item.update update the vector search index" and may require an "opt-in" approach for existing installs to do a crawl of the DB for generating this index. Some boundaries could also be that you may allow only certain collections (and I suppose forcibly their related collections) into the index, so as to limit usage to only necessary collections and reduce DB usage.

### Must Have:
- Implementation of semantic search methodology
- Controls/boundaries for result-sets ("always disallow show of items with 'hide' tag") and collection participation in semantic search.

### Should Have:
- Easy controls, as in not needing custom extensions/flows/flow operations to implement.

### Could Have:
- Vector search implementation that auto-updates on item.create/item.update
- Opt-in crawl of existing DB to generate vector index

### Won't Have:
- ?

### Drawbacks

- Potential for high DB usage (both in terms of compute and size)
- Potential for high complexity to code into Directus' architecture
- Could be too niche for implementation cost, but it is a "driving/complex feature" that can be advertised.

### Alternatives

Not much impact for not doing this.

Other designs would just be searching JSON fields and relational fields the same as we are "automatically" instead of doing complex filter rules that would change between pages within use-cases (ours for instance would take at least 5 different queries with deep/filter/etc in much complexity for having similar issues with word order).

### Adoption Strategy

Existing developers would probably have to opt-in for crawling the DB for a vector search index. It could potentially be a new query in the query parameters, or it could be a check for opt-in for ?search and then an "old" vs. "new" implementation depending on that opt-in (but then maintenance/improvements for two versions of search is required).

### Unresolved Questions

What methodology for semantic search? Vector search integration? How would we implement with existing search?

**Comment by br41nslug** [comment]
Closing this feature request in favor of https://github.com/directus/directus/discussions/17822 which addresses the same feature, has already had some discussion and has met the minimum amount of upvotes
